### ISSUE TITLE
Composed style takes precedence over specified style

### ISSUE THREAD

**Issue by wyqydsyq** [css-modules/css-modules#158]
I'm using PureCSS with CSS Modules and so far it's worked pretty well, however I've ran into an issue with the following code:

````
.submit {
	composes: pure-button from 'pure-css/lib/buttons.css';
	color: #f00;
}
````
The rendered element gets both relevant classes: `class="styles__submit___1xOVh buttons__pure-button___3gWmz"`

The issue I'm having is the `color` property from `.pure-button` is overriding the property I set on `.submit`, both classes have the same specificity and it seems the only way I can make the latter defined property take precedence is with `!important`.

I should be getting a button with red text however it's still the color inherited from `.pure-button`

**Comment by nkbt** [comment]
When you include css before component, overriding styles are placed before.
So we always include css after all other imports. I wish we have eslint
rule for that. Might be your case too

**Comment by geelen** [comment]
Yeah there's explicit test cases in there for this. If file `X` composes a class from file `Y` then `Y` is _guaranteed_ to be added to the source before `X`, so that if the two rules have the same specificity the files override in the right order.

Can you give us a bit more information? Are you using webpack/browserify/jspm? Any chance you can share the code or create a new project that isolates the bug?

**Comment by madpilot** [comment]
@geelen I've got one for you: https://github.com/madpilot/css-modules-bug-158

It looks like it happens when you compose from a file that has a compose.

If you look at output.css you can see the ```.include-2__header___2LJaB``` is *after* ```.include-1__header___H-qGE```

It should be the other way around.

**Comment by gamliela** [comment]
Hi,
Any update regarding this issue?

I created another [simple repo](https://github.com/gamliela/css-modules-test/) that demonstrates the problem. I use Webpack 2.2.

The root cause it that the style tag for the composed (base) class is injected twice, and the second time overrides the derived class.

Here are the results from Chrome Developer Tools:
```
.GuestList---1I1i- {
    color: blue;
}
.ModifiedGuestList---2Q8o_ {
    color: red;
}
.GuestList---1I1i- {
    color: blue;
}
```

**Comment by forestgagnon** [comment]
I've run into this issue with Sass files. My workaround has been to group overrides under a `&:not(:root)` selector, which always evaluates to true. I'd love to find a better solution.

e.g. 
```sass
.localClass
  composes: some-global-class from global
  &:not(:root)
    // Override properties from the global class
    width: 500px
    ...
```

**Comment by TrySound** [comment]
@michael-ciniawsky @evilebottnawi do you know this issue? Seams related to loaders.

**Comment by giannif** [comment]
I'm having the same issue and @gamliela's fix of using `importLoaders` didn't work, I was already using it

**Comment by gamliela** [comment]
I did find a fix, my problem was improper usage of `importLoaders`. Note that the order of loaders also matters. For a working version including SCSS check out my fix branch at repo https://github.com/gamliela/css-modules-test/

**Comment by giannif** [comment]
Thanks for the help @gamliela. I saw that fix branch, I have the same webpack config as you except instead of `sass-loader` I'm using `postcss-loader` and `css-modules`, and my config is identical to the [recommended setup](https://github.com/postcss/postcss-loader#css-modules) which includes `importLoaders`. 

The issue seems to be exactly what @madpilot has in https://github.com/madpilot/css-modules-bug-158. However, if I upgrade his `postcss-modules` dependency to the latest `0.8.0`, his issue goes away (he has a local path for his dep, but guessing it was around `0.5.0`)

Unfortunately, I'm not using `postcss-modules` since I'm using `css-loader`, `importLoaders` and `modules:true`. So I'm thinking my issue has to do with that combination 

Oddly enough, when I remove `importLoaders` the bug goes away. Will try to find time to dig in more

**Comment by dcousineau** [comment]
@giannif it's been almost a year, did you ever find a fix? I'm still chasing down this issue only in my project everything works when using style-loader in local dev but overrides breaks when i start using css-loader and an extract plugin.

**Comment by giannif** [comment]
@dcousineau no, I think I changed my css as a workaround, and the issue hasn't popped in other places, perhaps due to dependency updates. I also switched us to [styled-components](https://www.styled-components.com/) and no new code has used css-modules in about 6 months

**Comment by dcousineau** [comment]
Makes sense, I'm realizing I need to ditch `composes:` entirely as I've chased everything down to that being the culprit. Thanks!

**Comment by besserwisser** [comment]
I accepted to use !important to overwrite and will also use styled-components or other approaches in the next project.

**Comment by webdevbyjoss** [comment]
I've got the same issue reproduced recently and maybe it will be helpful for someone to take a look at the following statement:

https://github.com/css-modules/css-modules#composing-from-other-files


> Note that when composing multiple classes from different files the order of appliance is undefined. Make sure to not define different values for the same property in multiple class names from different files when they are composed in a single class.

This means that we should not expect the CSS properties to override when we are composing classes from different CSS files.

To get a predictable behavior we should avoid specifying the same CSS properties in the `submit` and `pure-button` classes.

To get the CSS properties to override, what we would expect intuitively from `composes:` property, we probably could use `!important`?

```css
.submit {
    composes: pure-button from 'pure-css/lib/buttons.css';
    color: #f00!important;
}
```

**Comment by raviqqe** [comment]
Is it too costly for CSS modules to define an implicit layer for each CSS module file? If that's possible, we can properly override properties defined in dependencies. We can infer layer names in the same way as class name prefices. For example,

`parent-a.module.css`:

```css
.parent {
  color: black;
}
```

`parent-b.module.css`:

```css
.parent {
  font-size: 16px;
}
```

`child.module.css`:

```css
.child {
  composes: parent from "./parent-a.css" parent from "./parent-b.css";
  color: red;
  font-size: 42px;
}
```

Those files are compiled into the following:

`parent-a.css`:

```css
@layer parent-b {
  .parent-a_parent {
    color: black;
  }
}
```

`parent-b.css`:

```css
@layer parent-b {
  .parent-b_parent {
    font-size: 16px;
  }
}
```

`child.css`:

```css
@layer parent-a, child;
@layer parent-b, child;

@layer child {
  .child_child {
    color: red;
    font-size: 42px;
  }
}
```

**Comment by BrendanBerkley** [comment]
I was looking at the @layer MDN docs, where they say that `@import "theme.css" layer(utilities);` is a thing. 

Could CSS Modules do something like this, and would it be a good idea?

`composes: buttonReset from "./layout.module.css" layer(resets);`

That way, layers can be a part of CSS Modules, but in a fully opt-in way.
